File: backend/routers/candles.py
"""
candles.py

REST endpoint for historical candles — the counterpart to routers/ws.py's
live push. Reads the Postgres serving copy (candles_1m) populated by
streaming/postgres_sink.py; deliberately never touches Redis, which only
ever holds the current candle and has no history to serve.

Bigger timeframes are rolled up from candles_1m at query time (see
postgres_client.py). `before` pages backwards for the chart's infinite
scroll-back: pass the window_start of the oldest candle already loaded.

Author: @DS
"""

from __future__ import annotations

from datetime import datetime, timedelta, timezone

from fastapi import APIRouter, HTTPException, Query

from postgres_client import CandleHistoryReader

router = APIRouter()

# TODO(@DS): move to config/settings.py once the config/ module is scaffolded.
DEFAULT_LIMIT = 200
# One day of 1m candles — what frontend/src/hooks/useTimeframeCandle.js
# fetches to seed a live 1D candle that's already in progress.
MAX_LIMIT = 1440

# Must match frontend/src/lib/timeframes.js's TIMEFRAMES.
TIMEFRAMES = {
    "1m": timedelta(minutes=1),
    "5m": timedelta(minutes=5),
    "15m": timedelta(minutes=15),
    "1h": timedelta(hours=1),
    "4h": timedelta(hours=4),
    "1D": timedelta(days=1),
}


@router.get("/candles/{symbol}")
def get_candle_history(
    symbol: str,
    timeframe: str = "1m",
    limit: int = Query(DEFAULT_LIMIT, ge=1, le=MAX_LIMIT),
    before: datetime | None = None,
) -> list[dict]:
    """Return up to `limit` candles for `symbol` at `timeframe`, oldest first.

    With `before`, only candles whose window starts before it.
    """
    bucket = TIMEFRAMES.get(timeframe)
    if bucket is None:
        raise HTTPException(status_code=422, detail=f"timeframe must be one of {list(TIMEFRAMES)}")
    if before is not None and before.tzinfo is None:
        before = before.replace(tzinfo=timezone.utc)  # a naive timestamp means UTC, like everything else here

    reader = CandleHistoryReader()
    try:
        return reader.get_candles(symbol, limit=limit, bucket=bucket, before=before)
    finally:
        reader.close()

File: frontend/src/api/candles.js
/**
 * candles.js
 *
 * The frontend's one client for backend/routers/candles.py's
 * GET /candles/{symbol} — shared by useCandleHistory.js (newest page),
 * CandleChart.jsx (older pages on scroll-back) and useTimeframeCandle.js
 * (seeding a live candle that's already in progress).
 *
 * Author: @DS
 */

// TODO(@DS): move to a config module once the frontend has one; also
// needs to become an env-driven URL before this is ever deployed anywhere
// but localhost.
const BACKEND_HTTP_URL = "http://localhost:8000";

// Candles per request. A page shorter than this means history has run out.
export const HISTORY_PAGE_SIZE = 300;

/**
 * Fetch candles for `symbol` at `timeframe`, oldest first. `before` (Unix
 * seconds) pages backwards: only candles starting before it are returned.
 */
export async function fetchCandles(symbol, { timeframe, limit = HISTORY_PAGE_SIZE, before } = {}) {
  const params = new URLSearchParams({ timeframe, limit: String(limit) });
  if (before !== undefined) params.set("before", new Date(before * 1000).toISOString());

  const response = await fetch(`${BACKEND_HTTP_URL}/candles/${encodeURIComponent(symbol)}?${params}`);
  if (!response.ok) {
    throw new Error(`GET /candles/${symbol} failed: ${response.status} ${response.statusText}`);
  }
  return response.json();
}
